lib/spack/spack/util/compression.py

```diff
@@ -59,21 +157,124 @@ def _unzip(archive_file):
     unzip = which(exe, required=True)
     unzip.add_default_arg(arg)
     unzip(archive_file)
+    return destination_abspath
+
+
+def _unZ(archive_file):
+    if is_windows:
+        result = _7zip(archive_file)
+    else:
+        result = _untar(archive_file)
+    return result
+
+
+def _lzma_decomp(archive_file):
+    """Decompress lzma compressed files. Prefer Python native
+    lzma module, but fall back on command line xz tooling
+    to find available Python support. This is the xz command
+    on Unix and 7z on Windows"""
+    if lzma_support():
+        import lzma  # novermin
+        _, ext = os.path.splitext(archive_file)
+        decompressed_file = os.path.basename(archive_file.strip(ext))
+        archive_out = os.path.join(os.getcwd(), decompressed_file)
+        with open(archive_out, 'wb') as ar:
+            with lzma.open(archive_file) as lar:
+                ar.write(lar.read())
+    else:
+        if is_windows:
+            return _7zip(archive_file)
+        else:
+            return _xz(archive_file)
+
+
+def _xz(archive_file):
+    """Decompress lzma compressed .xz files via xz command line
+    tool. Available only on Unix
+    """
+    if is_windows:
+        raise RuntimeError('XZ tool unavailable on Windows')
+    _, ext = os.path.splitext(archive_file)
+    outfile = os.path.basename(archive_file.strip(ext))
+    xz = which('xz', required=True)
+    xz.add_default_arg('-d')
+    xz(archive_file)
+    return outfile
+
+
+def _7zip(archive_file):
+    """Unpack/decompress with 7z executable
+    7z is able to handle a number file extensions however
+    it may not be available on system.
+
+    Without 7z, Windows users with certain versions of Python may
+    be unable to extract .xz files, and all Windows users will be unable
+    to extract .Z files. If we cannot find 7z either externally or a
+    Spack installed copy, we fail, but inform the user that 7z can
+    be installed via `spack install 7zip`
+
+    Args:
+        archive_file (str): absolute path of file to be unarchived
+    """
+    _, ext = os.path.splitext(archive_file)
+    outfile = os.path.basename(archive_file.strip(ext))
+    _7z = which('7z')
+    if not _7z:
+        raise CommandNotFoundError("7z unavailable,\
+unable to extract %s files. 7z can be installed via Spack" % ext)
+    _7z.add_default_arg('e')
+    _7z(archive_file)
+    return outfile
 
 
-def decompressor_for(path, extension=None):
-    """Get the appropriate decompressor for a path."""
-    if ((extension and re.match(r'\.?zip$', extension)) or
+def decompressor_for(path, ext=None):
+    """Wrapper for select_decompressor_for, returns
+    a function pointer to appropriate decompression
+    algorithm.
+
+    On Unix - simply invokes select_decompressor_for
+
+    On Windows - archives with one extension are
+    passed through to select_decompressor_for. Files
```

Comment:
This docstring looks out-of-date: I don't see `select_decompressor_for` anymore (nor composition of decompressors: i.e. `tar.gz` goes to `_untar`)